arangod/IResearch/IResearchDocument.cpp

```diff
@@ -715,6 +705,272 @@ bool StoredValue::write(irs::data_output& out) const {
+}
+
+bool InvertedIndexFieldIterator::setValue(VPackSlice const value,
+                             FieldMeta::Analyzer const& valueAnalyzer) {
+  TRI_ASSERT(value.isObject()
+            || value.isArray()
+            || value.isString());
+
+  auto& pool = valueAnalyzer._pool;
+
+  if (!pool) {
+    LOG_TOPIC("189db", WARN, iresearch::TOPIC)
+        << "got nullptr analyzer factory";
+
+    return false;
+  }
+
+  irs::string_ref valueRef;
+  AnalyzerValueType valueType{AnalyzerValueType::Undefined};
+
+  switch (value.type()) {
+    case VPackValueType::Array: {
+      valueRef = iresearch::ref<char>(value);
+      valueType = AnalyzerValueType::Array;
+    } break;
+    case VPackValueType::Object: {
+      valueRef = iresearch::ref<char>(value);
+      valueType = AnalyzerValueType::Object;
+    } break;
+    case VPackValueType::String: {
+      valueRef = iresearch::getStringRef(value);
+      valueType = AnalyzerValueType::String;
+    } break;
+    default:
+      TRI_ASSERT(false);
+      return false;
+  }
+
+  if (!pool->accepts(valueType)) {
+    return false;
+  }
+
+  // init stream
+  auto analyzer = pool->get();
+
+  if (!analyzer) {
+    LOG_TOPIC("22eeb", WARN, arangodb::iresearch::TOPIC)
+        << "got nullptr from analyzer factory, name '" << pool->name() << "'";
+    return false;
+  }
+  if (!analyzer->reset(valueRef)) {
+      return false;
+  }
+  // set field properties
+  switch (pool->returnType()) {
+    case AnalyzerValueType::Bool:
+      {
+        if (!analyzer->next()) {
+          return false;
+        }
+        _currentTypedAnalyzer = std::move(analyzer);
+        _currentTypedAnalyzerValue = irs::get<VPackTermAttribute>(*analyzer);
+        TRI_ASSERT(_currentTypedAnalyzerValue);
+        setBoolValue(_currentTypedAnalyzerValue->value);
+        _primitiveTypeResetter = [](irs::token_stream* stream, VPackSlice slice) -> void {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+          auto bool_stream = dynamic_cast<irs::boolean_token_stream*>(stream);
+          TRI_ASSERT(bool_stream);
+#else
+          auto bool_stream = static_cast<irs::boolean_token_stream*>(stream);
+#endif
+          TRI_ASSERT(slice.isBool());
+          bool_stream->reset(slice.getBool());
+        };
+      }
+      break;
+    case AnalyzerValueType::Number:
+      {
+        if (!analyzer->next()) {
+          return false;
+        }
+        _currentTypedAnalyzer = std::move(analyzer);
+        _currentTypedAnalyzerValue = irs::get<VPackTermAttribute>(*analyzer);
+        TRI_ASSERT(_currentTypedAnalyzerValue);
+        setNumericValue(_currentTypedAnalyzerValue->value);
+        _primitiveTypeResetter = [](irs::token_stream* stream, VPackSlice slice) -> void {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+          auto number_stream = dynamic_cast<irs::numeric_token_stream*>(stream);
+          TRI_ASSERT(number_stream);
+#else
+          auto number_stream = static_cast<irs::numeric_token_stream*>(stream);
+#endif
```

Comment:
Maybe you should write a macro for such a pattern?
Better yet, a `constexpr` function

```C++
template <typename To, typename From>
constexpr ptrCast(From&& from) {
#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
  auto* to = dynamic_cast<To*>(std::forward<From>(from));
  TRI_ASSERT(to);
  return to;
#else
  return static_cast<To*>(std::forward<From>(from));
#endif
}
```

Reply:
We decide make separate PR for this